### ISSUE TITLE
bench: extend benchmark/run.py to detect the perf-issue classes in #101-104

### ISSUE THREAD

**Issue by mmcdermott** [mmcdermott/meds-torch-data#105]
## Summary

The existing `benchmark/run.py` runs a single end-to-end pass on a fixed config and reports aggregate throughput. It catches catastrophic regressions but doesn't diagnose the perf-issue classes surfaced in [#101](https://github.com/mmcdermott/meds-torch-data/issues/101) (per-getitem NRT open overhead), [#102](https://github.com/mmcdermott/meds-torch-data/issues/102) (DataLoader defaults), [#103](https://github.com/mmcdermott/meds-torch-data/issues/103) (schema duplication across workers), or [#104](https://github.com/mmcdermott/meds-torch-data/issues/104) (skewed-length padding). Extending the benchmark to exercise those code paths and report signal-specific metrics would let PRs that regress any one of them surface the issue on the github-benchmark-action comment.

## Current state

[`benchmark/run.py`](https://github.com/mmcdermott/meds-torch-data/blob/main/benchmark/run.py): one `test_profile` parametrized over batch_size / max_seq_len / num_epochs, wrapping a full DataLoader iteration. Uses `MemTrackableMixin` and `TimeableMixin` to track time + peak RSS for the whole run. Result feeds [`github-action-benchmark`](https://github.com/benchmark-action/github-action-benchmark) for per-PR regression alerts.

Good as a coarse regression detector. Limited for diagnosis because:

1. Everything is measured at the *run* level. A regression in `__getitem__` vs `collate` vs schema loading all look the same.
2. Uses whatever the builder wires for DataLoader settings — doesn't vary `num_workers` / `pin_memory` / `persistent_workers` explicitly.
3. Fixed cohort shape (the existing toy fixture) — doesn't capture skew or scale effects.

## Proposal

Add targeted benchmarks alongside the existing end-to-end one. Each should run cheaply enough (≤ 30s) to fit inside the existing Benchmark Performance workflow budget and emit its own metric that github-action-benchmark tracks independently.

### (1) `__getitem__` microbenchmark → detects #101

- Iterate `dataset[i]` for i in range(N), no DataLoader, no collate.
- Report mean / p50 / p95 wall time per call.
- Regression signal: if NRT handle caching regresses or someone reintroduces unnecessary work in `load_subject_data`, this metric rises sharply.
- Tiny cost: 1000 calls × 300 µs ≈ 300 ms per benchmark run.

### (2) DataLoader config sweep → detects #102

- Sweep `num_workers ∈ {0, 2, 4}` × `pin_memory ∈ {False, True}` × `persistent_workers ∈ {False, True}`.
- Report samples/sec for each cell on a fixed 1000-item pass.
- Regression signal: the `(4, True, True)` cell should stay ≥ some multiple of `(0, False, False)`. If they converge, something broke the worker parallelism.
- Cost: 12 cells × ~5s each ≈ 1 min.

### (3) Larger-scale synthetic cohort → detects #103

- Generate a 10k-subject synthetic cohort (scale up the existing fixture builder with an `n_subjects` parameter).
- Report peak RSS per-worker × `num_workers ∈ {0, 2, 4, 8}`.
- Regression signal: if schema representation grows or `persistent_workers=True` stops being used, per-worker RSS rises.
- Cost: one-time fixture build (cached across runs), then ~30s of iteration at each num_workers setting.

### (4) Skewed-length synthetic cohort → detects #104

- Generate a cohort where `events_per_subject` follows Zipf(1.5) or similar long-tail distribution.
- Measure batch-construction time and peak allocation with/without a length-bucketed sampler.
- Regression signal: if `max_seq_len` truncation regresses or `to_dense` behavior changes upstream, the skewed variant's time diverges from the uniform variant's.
- Cost: two benchmark runs, ~30s each.

## Implementation sketch

Keep each new benchmark in its own file under `benchmark/` with the existing `MemTrackableMixin`/`TimeableMixin` pattern so they hook into the same github-action-benchmark pipeline that `run.py` already feeds. Emit a distinct benchmark name per metric so each shows up as its own tracked series in the performance-alert comment.

```
benchmark/
├── run.py               # existing end-to-end
├── getitem_latency.py   # (1)
├── dataloader_sweep.py  # (2)
├── schema_memory.py     # (3)
└── skewed_lengths.py    # (4)
```

## Upstream factors

- **github-action-benchmark** supports multiple benchmark series per run; no config change beyond what we already have.
- **`meds_testing_helpers`**: the synthetic-cohort builder used by our test fixtures is there. Scaling it up to 10k subjects is a parameter sweep, not a rewrite — but the exact API depends on what the builder exposes. May need an upstream issue there if the knob isn't there yet.
- **`ml-mixins` (`MemTrackableMixin` / `TimeableMixin`)**: already pulled in for `benchmark/`. Should be enough for (1)-(4).

## Related

- [#101](https://github.com/mmcdermott/meds-torch-data/issues/101) — NRT handle caching (detected by (1))
- [#102](https://github.com/mmcdermott/meds-torch-data/issues/102) — Datamodule defaults (detected by (2))
- [#103](https://github.com/mmcdermott/meds-torch-data/issues/103) — schema RSS duplication (detected by (3))
- [#104](https://github.com/mmcdermott/meds-torch-data/issues/104) — padding / skewed lengths (detected by (4))

**Comment by mmcdermott** [comment]
See some of the comments in https://github.com/mmcdermott/nested_ragged_tensors/issues/55 for thoughts on improvements here.

**Comment by mmcdermott** [comment]
Good pointer — a lot of mmcdermott/nested_ragged_tensors#55 applies here, and some of it should drive how #105 is structured rather than being orthogonal. Specifics worth adopting:

**Structural alignment with NRT#55:**

- **Tiered rollout** (NRT#55's Essential / Valuable / Nice split). #105 currently lists four benchmarks as equal peers. Better ordering for us:
  - *Tier 1 (Essential)*: `getitem_latency.py` (1) and phase-splitting the existing `run.py` epoch into `getitem_total` / `collate_total` / `dataloader_overhead`. Cheap, high signal, required for any of the #101–#104 fixes to have a landing zone.
  - *Tier 2 (Valuable)*: `dataloader_sweep.py` (2), `schema_memory.py` (3) — these need a larger synthetic cohort fixture (a genuine prereq, not a trivial add).
  - *Tier 3 (Conditional)*: `skewed_lengths.py` (4) — only worth building if #9 isn't addressed in the near term, since #9's acceptance test would fold this work in (per my reply on #104).

- **Phase splitting inside `run.py`** (NRT#55 item B) is actually more valuable than I gave it credit for. A bench that outputs `getitem_total`, `collate_total`, `dataloader_overhead` separately diagnoses #101 vs #102 vs #104 regressions *in the existing end-to-end run* without needing the dedicated micro-bench first. That's the single highest-signal-per-LoC change in this whole list.

- **Naive-padding baseline** (NRT#55 item D): would give absolute context for #104's skew problem vs padded-dense. Worth doing at the NRT layer only — replicating it at MTD is redundant.

**Where MTD diverges from NRT#55:**

- **Tooling choice**. NRT#55 proposes `pytest-benchmark`; MTD currently uses `MemTrackableMixin` / `TimeableMixin` piped through github-action-benchmark. Keep the current approach — the action's `customSmallerIsBetter` output is already wired end-to-end, and `pytest-benchmark`'s JSON doesn't round-trip cleanly into it. If NRT#55 migrates, we revisit.

- **Memory tracking is an MTD-specific need**. `schema_memory.py` (3) is about RSS-per-worker, which NRT#55 doesn't cover because NRT isn't worker-aware. Keep as-is, distinct from upstream.

- **GH Pages display improvements** (NRT#55 Tier 3: grouping, summary tables, std-dev surfacing). These apply verbatim to MTD's benchmark page too — worth a separate sub-task here. Especially the markdown-summary-in-CI-job-output idea; a table with "metric / value / Δ vs. prev" would be more useful than the current flat chart list.

**Revised plan for #105:**

1. Phase-split `run.py` (cheap, high signal) — do this first.
2. Add `getitem_latency.py` — regression anchor for #101.
3. Parameterize the fixture for scale, then add `dataloader_sweep.py` + `schema_memory.py`.
4. Defer `skewed_lengths.py` — roll into #9's rollout plan instead.
5. GH Pages grouping + CI markdown-summary-table as a separate display-focused sub-task.

Want me to restructure the issue body to reflect this tiering and carve out the display work as a sub-issue?
